flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/internal/TransactionOwnership.java

```diff
@@ -0,0 +1,205 @@
+     *
+     * <p>With these assumption, this ownership model is able to recover from writer states with
+     * subtask id + max parallelism:
+     *
+     * <ul>
+     *   <li>If there is state, we can extract the owned subtask ids and the max parallelism from
+     *       the state.
+     *   <li>If there is no state, we can use the current subtask id and the max parallelism from
+     *       the current parallelism. The current subtask id cannot possibly be owned already. The
+     *       max parallelism in any state must be lower than the current max parallelism.
+     *   <li>Hence, no subtask id is owned by more than one task and all tasks have the same max
+     *       parallelism.
+     *   <li>Since all tasks have shared knowledge, we can exclusively assign all transactional ids.
+     *   <li>Since each subtask owns at least one transactional id, we can safely create new
+     *       transactional ids while other subtasks are still aborting their transactions.
+     * </ul>
+     */
+    EXPLICIT_BY_WRITER_STATE {
+        @Override
+        public int[] getOwnedSubtaskIds(
+                int currentSubtaskId,
+                int currentParallelism,
+                Collection<KafkaWriterState> recoveredStates) {
+            if (recoveredStates.isEmpty()) {
+                return new int[] {currentSubtaskId};
+            } else {
+                int[] ownedSubtaskIds =
+                        recoveredStates.stream()
+                                .mapToInt(KafkaWriterState::getOwnedSubtaskId)
+                                .sorted()
+                                .toArray();
+                assertKnown(ownedSubtaskIds[0]);
+
+                int maxParallelism =
+                        recoveredStates.iterator().next().getTotalNumberOfOwnedSubtasks();
+                // Assumption of the ownership model: state is distributed consecutively across the
+                // subtasks starting with subtask 0
+                checkState(currentSubtaskId < maxParallelism, "State not consecutively assigned");
+
+                return ownedSubtaskIds;
+            }
+        }
+
+        @Override
+        public int getTotalNumberOfOwnedSubtasks(
+                int currentSubtaskId,
+                int currentParallelism,
+                Collection<KafkaWriterState> recoveredStates) {
+            if (recoveredStates.isEmpty()) {
+                return currentParallelism;
+            }
+            int totalNumberOfOwnedSubtasks =
+                    recoveredStates.iterator().next().getTotalNumberOfOwnedSubtasks();
+            assertKnown(totalNumberOfOwnedSubtasks);
+
+            if (currentParallelism >= totalNumberOfOwnedSubtasks) {
+                // Assumption of the ownership model: state is distributed consecutively across the
+                // subtasks starting with subtask 0
+                checkState(recoveredStates.size() == 1, "Not uniformly assigned");
```

Comment:
Is this an upscale scenario? 

Shouldn't the state assignment be 0 or 1?

Reply:
Yes that's correct. But the case 0 cannot happen because of `if (recoveredStates.isEmpty())`